### PR TITLE
tests: Replaces mounttest images used with agnhost (part 4)

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind feature

/sig testing
/sig windows
/area conformance

**What this PR does / why we need it**:

Quite a few images are only used a few times in a few tests. Thus, the images are being centralized into the agnhost image, reducing the number of images that have to be pulled and used.

This PR replaces the usage of the following images with agnhost:

- mounttest
- mounttest-user

Additionally, removes the usage of the ``mounttest-user`` image and removes it from ``kubernetes/test/images``. ``RunAsUser`` is set instead of having that image.

**Which issue(s) this PR fixes**:

Related: #76342

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by smourapina** [comment]
Hello @PatrickLang, @claudiubelu!
This is Bug Triage for release 1.18 checking in, with a friendly reminder that code freeze is coming in about 1 week (5 March). Is this PR still intended for milestone 1.18?

File: test/e2e/common/empty_dir.go [hunk 1]
```diff
@@ -35,8 +35,8 @@ const (
 )
 
 var (
-	testImageRootUid    = imageutils.GetE2EImage(imageutils.Mounttest)
-	testImageNonRootUid = imageutils.GetE2EImage(imageutils.MounttestUser)
+	testImageRootUid = imageutils.GetE2EImage(imageutils.Agnhost)
+	nonRootUid       = int64(1001)
```
Comment (jingxu97) [inline test/e2e/common/empty_dir.go]: what is the reason for this change?

File: test/e2e/common/empty_dir.go [hunk 2]
```diff
@@ -50,27 +50,27 @@ var _ = ginkgo.Describe("[sig-storage] EmptyDir volumes", func() {
 		})
 
 		ginkgo.It("new files should be created with FSGroup ownership when container is root", func() {
-			doTestSetgidFSGroup(f, testImageRootUid, v1.StorageMediumMemory)
+			doTestSetgidFSGroup(f, nil, v1.StorageMediumMemory)
```
Comment (jingxu97) [inline test/e2e/common/empty_dir.go]: why using nil here?

File: test/e2e/common/empty_dir.go (hunk 1 above)
Comment (claudiubelu) [reply test/e2e/common/empty_dir.go]: So, we have decided to centralize some images into the agnhost image, and to remove images that are no longer very useful. See entire thread here: https://github.com/kubernetes/kubernetes/issues/76342

Among these images is ``mounttest-user``. Looking through the git history, this image was introduced a loooong time ago, before ``PodSecurityPolicy`` was a thing, and thus, you couldn't create pods and tell them to run as certain users; that had to be done in the image itself, and that's exactly what ``mounttest-user`` image's purpose was.

Nowadays, that is no longer needed, we can do that programatically, as you already know. As for why ``nonRootUid`` is ``1001``, that's because the ``mounttest-user`` image was running as UID ``1001``, so I'm just maintaining the behaviour.

File: test/e2e/common/empty_dir.go (hunk 2 above)
Comment (claudiubelu) [reply test/e2e/common/empty_dir.go]: That parameter changed from a string (specifying the image name) to to a *int64 (specifying the user as whom to run). ``nil`` besically means that it will run as the image's default user, which maintains the same behaviour.

**Comment by claudiubelu** [comment]
> 
> 
> Hello @PatrickLang, @claudiubelu!
> This is Bug Triage for release 1.18 checking in, with a friendly reminder that code freeze is coming in about 1 week (5 March). Is this PR still intended for milestone 1.18?

It is. The original PR was split into 3 parts, and one of the parts merged. This should too.

**Comment by PatrickLang** [comment]
@jingxu97 - Claudiu responded to the comments above - can you approve for SIG-Storage?

**Comment by markyjackson-taulia** [comment]
@jingxu97 gentle ping. Code freeze is this Thursday

**Comment by markyjackson-taulia** [comment]
@jingxu97 is this good to go? Code freeze is tomorrow. We either need to punt it or ask for an exception. I am on the West Coast and tomorrow morning, if there is no response I will have to move this out of the milestone.
Cc @claudiubelu

File: test/e2e/common/empty_dir.go [hunk 3]
```diff
@@ -301,14 +301,15 @@ const (
 	volumeName    = "test-volume"
 )
 
-func doTestSetgidFSGroup(f *framework.Framework, image string, medium v1.StorageMedium) {
+func doTestSetgidFSGroup(f *framework.Framework, uid *int64, medium v1.StorageMedium) {
 	var (
 		filePath = path.Join(volumePath, "test-file")
 		source   = &v1.EmptyDirVolumeSource{Medium: medium}
-		pod      = testPodWithVolume(testImageRootUid, volumePath, source)
+		pod      = testPodWithVolume(uid, volumePath, source)
```
Comment (jingxu97) [inline test/e2e/common/empty_dir.go]: strange thing is in the previous code, the passed image is not used. it always use testImageRootUid. That means testImageNonRootUid image is not tested at all?

File: test/e2e/common/empty_dir.go [hunk 4]
```diff
@@ -35,8 +35,8 @@ const (
 )
 
 var (
-	testImageRootUid    = imageutils.GetE2EImage(imageutils.Mounttest)
-	testImageNonRootUid = imageutils.GetE2EImage(imageutils.MounttestUser)
+	testImageRootUid = imageutils.GetE2EImage(imageutils.Agnhost)
```
Comment (jingxu97) [inline test/e2e/common/empty_dir.go]: maybe change the name of the image? It sounds like a uid instead of image name.

File: test/e2e/common/empty_dir.go [hunk 5]
```diff
@@ -151,7 +151,7 @@ var _ = ginkgo.Describe("[sig-storage] EmptyDir volumes", func() {
 		This test is marked LinuxOnly since Windows does not support setting specific file permissions.
 	*/
 	framework.ConformanceIt("volume on default medium should have the correct mode [LinuxOnly] [NodeConformance]", func() {
-		doTestVolumeMode(f, testImageRootUid, v1.StorageMediumDefault)
+		doTestVolumeMode(f, nil, v1.StorageMediumDefault)
```
Comment (jingxu97) [inline test/e2e/common/empty_dir.go]: just to make it clear, maybe add comments after nil /*root user*/?

File: test/e2e/storage/testsuites/subpath.go [hunk 6]
```diff
@@ -614,8 +618,8 @@ func SubpathTestPod(f *framework.Framework, subpath, volumeType string, source *
 }
 
 func containerIsUnused(container *v1.Container) bool {
-	// mountImage with nil Args does nothing. Leave everything else
-	return container.Image == mountImage && container.Args == nil
+	// mountImage with nil Args or with just "mounttest" as Args does nothing. Leave everything else
```
Comment (jingxu97) [inline test/e2e/storage/testsuites/subpath.go]: I see this function is used removeUnusedContainers(). do you know what is the purpose of doing it?

**Comment by jingxu97** [comment]
just a couple of small things. Otherwise lgtm.

File: test/e2e/common/empty_dir.go (hunk 5 above)
Comment (claudiubelu) [reply test/e2e/common/empty_dir.go]: Well. I've updated this a bit. I've changed it from a pointer to just a regular ``int64``. Now, it's either ``0`` or ``nonRootUid``, which makes it a bit more clear, and it's also consistent with the other Volume related tests, which treat the root / non-root tests in a similar fashion.

Additionally, I've added a comment to ``testPodWithVolume`` (which is actually called by all the tests) mentioning that if the UID is 0, the Pod will run with with its default user (``root``, or if we'll add Windows support one day, ``ContainerAdministrator``)

File: test/e2e/common/empty_dir.go (hunk 3 above)
Comment (claudiubelu) [reply test/e2e/common/empty_dir.go]: Indeed, this function ignores the given ``image``, which is a bug. But in other functions, like ``doTestSubPathFSGroup``, it uses the given ``image``, ``testImageNonRootUid`` was used by other tests.

File: test/e2e/storage/testsuites/subpath.go (hunk 6 above)
Comment (claudiubelu) [reply test/e2e/storage/testsuites/subpath.go]: Reading through the comments, it seems that there are tests that are spawning pods with multiple containers, and it's expecting some of the containers to fail (negative testing). In that case, it removes those failed containers, as well as any init containers.

**Comment by matthyx** [comment]
I guess we've missed the 1.18 release window...
/lgtm

**Comment by markyjackson-taulia** [comment]
@claudiubelu @matthyx @jingxu97 
This is not going to make it in milestone v1.18. We are past the point of exception. I will move to the v1.19 milestone.

**Comment by claudiubelu** [comment]
So, can we proceed with this PR? It's been up for quite a while, even though it was supposed to merge ASAP. It also has a lgtm.
